### PR TITLE
Fix Singletons and Singleton Spawn

### LINKED ISSUE #10590: Singleton MPI initialization and spawn
Looks like singleton MPI init and spawn is broken in the main branch.

Look at this reproducer:

```c
#include <mpi.h>
#include <stdio.h>

int main(int argc, char *argv[])
{
  MPI_Comm parent, intercomm;

  MPI_Init(NULL, NULL);

  MPI_Comm_get_parent(&parent);
  if (MPI_COMM_NULL != parent)
    MPI_Comm_disconnect(&parent);
  
  if (argc > 1) {
    printf("Spawning '%s' ... ", argv[1]);
    MPI_Comm_spawn(argv[1], MPI_ARGV_NULL,
                   1, MPI_INFO_NULL, 0, MPI_COMM_SELF,
                   &intercomm, MPI_ERRCODES_IGNORE);
    MPI_Comm_disconnect(&intercomm);
    printf("OK\n");
  }

  MPI_Finalize();
}
```

Now I run that code using Open MPI v4.1.2 (system package from Fedora 36) the following two ways:

```
$ mpiexec -n 1 ./a.out ./a.out 
Spawning './a.out' ... OK

$ ./a.out ./a.out 
Spawning './a.out' ... OK
```

Note that the second way does not use `mpiexec` (that is, what the MPI standard calls singleton MPI initialization).

Next I run the code with ompi/main. I've configured with:

```
./configure \
    --without-ofi \
    --without-ucx \
    --with-pmix=internal \
    --with-prrte=internal \
    --with-libevent=internal \
    --with-hwloc=internal \
    --enable-debug \
    --enable-mem-debug \
    --disable-man-pages \
    --disable-sphinx
```

The first way (using `mpiexec`) seems to works just fine. The second way (singleton MPI init) fails:

```
$ mpiexec -n 1 ./a.out ./a.out 
Spawning './a.out' ... OK

$ ./a.out ./a.out 
[kw61149:1105609] OPAL ERROR: Error in file ../../ompi/dpm/dpm.c at line 2122
[kw61149:00000] *** An error occurred in MPI_Comm_spawn
[kw61149:00000] *** reported by process [440139776,0]
[kw61149:00000] *** on communicator MPI_COMM_SELF
[kw61149:00000] *** MPI_ERR_UNKNOWN: unknown error
[kw61149:00000] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[kw61149:00000] ***    and MPI will try to terminate your MPI job as well)
```

PS: Lack of singleton MPI initialization complicate some Python users wanting to dynamically spawn MPI processes as needed via mpi4py without requiring the parent process to be launched through `mpiexec`.

### PR DESCRIPTION
**Pull request by jjhursey**
* Fixes #10590
 * Singletons will not have a PMIx value for `PMIX_LOCAL_PEERS`
   so make that optional instead of required.
 * `&` is being confused as an application argument in `prte`
   instead of the background character
   * Replace with `--daemonize` which is probably better anyway

### CODE REVIEW

**Comment by jjhursey** [comment]
@awlauria We will need to sync the prrte submodule pointer to pick up https://github.com/openpmix/prrte/pull/1443

**Comment by rhc54** [comment]
FWIW: I have fixed the `&` confusion in the PMIx command line parser in https://github.com/openpmix/openpmix/pull/2694. You need to update the PMIx submodule pointer once that has been committed, and add https://github.com/open-mpi/ompi/pull/10695, to fix https://github.com/open-mpi/ompi/issues/10691.

What a tangled web we weave!

**Comment by jsquyres** [comment]
@jjhursey When singletons are fully fixed, please merge https://github.com/open-mpi/ompi-scripts/pull/62 so that singleton tests are added to the OMPI Jenkins CI.

**Comment by jjhursey** [comment]
Hold this PR for combined testing. I'm trying to align the 3 repos to get a current view of the state of this work.

I'm testing with Open MPI `main`:
 * OMPI PRs:
   * https://github.com/open-mpi/ompi/pull/10688
   * https://github.com/open-mpi/ompi/pull/10695
 * Submodules (HEAD of OpenPMIx v4.2.0 and PRRTE v3.0.0)
```
shell$ git submodule status
+8cb6f58fe074efde0239aa4567854742223ab1a9 3rd-party/openpmix (v4.2.0-3-g8cb6f58f)
+b29abde61c618de28f6a4c181e8c28f68e332969 3rd-party/prrte (v3.0.0rc1-18-gb29abde6)
```

With these changes it seems the things are failing again :( I'm investigating.

**Comment by jjhursey** [comment]
Tested with Open MPI `main`
 * OMPI PRs:
   * https://github.com/open-mpi/ompi/pull/10688
   * https://github.com/open-mpi/ompi/pull/10695
 * Submodules (HEAD of OpenPMIx master and PRRTE master)
```
shell$ git submodule status
+e3b925f82d2a59f58c60c7d7a7b5a71eda7d41ae ../../3rd-party/openpmix (v1.1.3-3598-ge3b925f8)
+12bb6c7dd6df522a38ed611c1fa4cf2dc9ea1761 ../../3rd-party/prrte (psrvr-v2.0.0rc1-4410-g12bb6c7d)
```

So something must be missing from the OpenPMIx and/or PRRTE release branches.

For Open MPI, since it uses the `master` branch of those two projects this PR is fine to merge.

**Comment by jjhursey** [comment]
FYI: I pushed my set of tests https://github.com/open-mpi/ompi-tests-public/pull/20

**Comment by jjhursey** [comment]
* :white_check_mark: Open MPI `main` with OpenPMIx `master` and PRRTE `master` works.
 * :boom: Open MPI `main` with OpenPMIx `v4.2` and PRRTE `v3.0` does not work.
 * :boom: Open MPI `main` with OpenPMIx `v4.2` and PRRTE `master` does not work
 * :white_check_mark: Open MPI `main` with OpenPMIx `master` and PRRTE `v3.0` does work

So it looks like an OpenPMIx issue - probably a missing commit from `master`. 👀

**Comment by jjhursey** [comment]
Ok we resolved the issue with OpenPMIx `v4.2` as recommended by https://github.com/openpmix/openpmix/issues/2705#issuecomment-1222904855 . I posted https://github.com/open-mpi/ompi/pull/10700 with the change.

Once Open MPI `main` has the following merged then spawn should work correctly.
 * https://github.com/open-mpi/ompi/pull/10695
 * https://github.com/open-mpi/ompi/pull/10688
 * https://github.com/open-mpi/ompi/pull/10700
 * https://github.com/open-mpi/ompi/pull/10699

Once verified then we will PR these back to Open MPI v5.x

**Review by awlauria — APPROVED** [review]

**Pull request merged**
